src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -511,6 +515,96 @@ private static ParsingStatus HexNumberToBigInteger(ref BigNumberBuffer number, o
             }
         }
 
+        private static ParsingStatus BinNumberToBigInteger(ref BigNumberBuffer number, out BigInteger result)
+        {
+            if (number.digits is null || number.digits.Length < 2)
+            {
+                result = default;
+                return ParsingStatus.Failed;
+            }
+
+            const int DigitsPerBlock = 32;
+
+            int totalDigitCount = number.digits.Length - 1;   // Ignore trailing '\0'
+
+            int blockCount = Math.DivRem(totalDigitCount, DigitsPerBlock, out int remainingDigitsInBlock);
+            if (remainingDigitsInBlock == 0)
+            {
+                remainingDigitsInBlock = DigitsPerBlock;
+            }
+            else
+            {
+                blockCount++;
+            }
+
+            Debug.Assert(number.digits[0] is '0' or '1');
+            bool isNegative = number.digits[0] == '1';
+
+            uint[]? arrayFromPool = null;
+            Span<uint> bufferSpan = blockCount <= BigIntegerCalculator.StackAllocThreshold ?
+                stackalloc uint[blockCount] : (arrayFromPool = ArrayPool<uint>.Shared.Rent(blockCount)).AsSpan(0, blockCount);
+
+            try
+            {
+                uint currentBlock = isNegative ? 0xFF_FF_FF_FF : 0x0;
+                int bufferPos = blockCount;
+                foreach (ReadOnlyMemory<char> chunkMem in number.digits.GetChunks())
```

Comment:
Having them be similar gives us more opportunities to deduplicate the code later and makes it easier to compare/contrast for bugs or other perf improvements.

Reply:
Changed code in `BinNumberToBigInteger()` to maximize matching `HexNumberToBigInteger()` general pattern.

BTW, for another direction, also changed some code in `FormatBigIntegerToBin()` to match some pattern of `FormatBigIntegerToHex` but I would not align other parts for it because I feel it is not unclear deviation from existing `FormatBigIntegerToHex()`. Brief reason explained [here](https://github.com/dotnet/runtime/pull/85392#discussion_r1213265779), sorry..

Please check, thanks @tannergooding !